Feature: individual import allows location and checks user's row-level security

weilu opened the pull request:
- [x] validate_import_individuals (used by DataUploadWorkflow and DataUpdateWorkflow) --- services.py
- [x] import_individuals ---- views.py
- [x] download_template_file
- [x] download_invalid_items
- [x] download_individual_upload
- [x] base_individual_upload --- workflows
- [x] individual_upload_valid 
- [x] base_individual_update
- [x] individual_update_valid 

In the process of working on this feature I added tests to cover the existing functionalities of the template download, import, and import validation. I also came across some previous refactoring – I believe to address the performance issue of individual uniqueness validation – that is incomplete, as evidenced by the leftover unused code. There was also a bug uncovered by the added tests, so I'm putting up this draft PR for quick review by @sniedzielski first of the refactored change and bug fix.

Also, @sniedzielski, not sure if you are aware, the with code before this PR the uniqueness check doesn't use the code in calcrule_validations/strategies/deduplication_individual_validation_strategy.py anymore, to which you are the original author and instead the uniqueness validation result is simply `national_id_uniqueness: false`. This PR puts back the dictionary structure but doesn't put back the attribute `duplications` as it would involve lots of database calls (which I believe was where the performance optimization your teammate did), so now with the PR, the validation result becomes 
```
"national_id_uniqueness":{
               "success": false,
               "field_name": "national_id",
               "note": "'national_id' Field value '1345320000AN' is duplicated",
}
```
Can you have a look at the relevant changes in this PR and let me know if it's ok? If so, I'll also go ahead and update the social_protection module's README to remove the duplications sample output to keep things consistent: https://github.com/openimis/openimis-be-social_protection_py/blob/b7c5359cf78898c061c7482bddff3c273adbf96f/README.md?plain=1#L202-L274

Should deduplication_individual_validation_strategy.py be removed from calcrule_validations if it's not used anymore?

weilu committed:
Add location_name to individuals import template

weilu committed:
View tests for import individuals endpoint

weilu committed:
Remove unused view function, variable and imports

weilu committed:
Happy path test case for IndividualImportService’s import_individuals

weilu committed:
Fix uniqueness validation result recording

Previous implementation would cause an error in 
save_validation_error_in_data_source_bulk’s value.get('success', False)
when uniqueness check is present

weilu committed:
Remove unused validation strategy config

sniedzielski commented:
@weilu Thanks for fixing the bugs and adding tests. To be honest, the uniqueness check in the validation module seems like overkill - of course because of our initial plan. It doesn't pass our expectation therefore it's good opportunity to rid out of it now. Your suggestion to simplify it is spot on for improving performance and clarity. I agree with removing the relevant README sections and handling uniqueness in the validation as you suggested in your initial comment on the PR. Please go ahead and continue with that. 
Just let me know if you have final version.

weilu committed:
Add row-level security to individual import validation

weilu committed:
Workflow dataframe headers validation allows location_name

weilu committed:
Rename view_test to test_view 

so it gets picked up automatically

weilu committed:
Basic scenarios for testing process_import_individuals_workflow

weilu committed:
Allow empty location to pass validation for individuals upload

weilu committed:
Individual upload workflow handles location

weilu committed:
Basic tests for workflow individual update

Also modify where config is patched for individual upload test

weilu committed:
Individual import update workflow handles location

weilu committed:
Individuals import with location will also assign location to group

Group takes on head’s location when it doesn’t have one, and
group member assignment results in alignment with group location

weilu committed:
Group update triggers location alignment among its members

weilu committed:
fake_individuals command also generate location_name column & values

weilu committed:
Individual import checks that location name is not ambiguous against DB

weilu committed:
Conditionally skip tests whose implementation uses a Postgres extension

weilu committed:
Use both location name and code for individual upload

To minimize ambiguous location issue

weilu committed:
Filter all location queries by validity

weilu commented:
Closing this as the same change set is included and merged in #137
